refactor!: splitting `getLogsByTags` to private and public versions

benesjan opened the pull request:
Fixes #14555

## New PR Description

As mentioned in the old PR description I originally planned to just type the arg of `getLogsByTags` but while addressing this PR's feedback I realized that I made a mistake: the `getLogsByTags` endpoint was used for both private and public logs and the issue is that for public logs we use "unsiloed" tag and for private logs we use siloed tag.

Having these 2 endpoints merged was just a tech debt: there is not a place where we would call this endpoint expecting only 1 type of logs. This led to us having ugly utilities that filtered out one kind of logs after retrieval.

We didn't split this before because we didn't want to introduce a breaking change to the API.

Given that I am already breaking backwards compatibility and given the error I made it made sense to fix this by splitting the two endpoints.

Now we have `getPrivateLogsByTag` and `getPublicLogsByTagsFromContract` endpoints.

## Old PR Description
In this PR I used the `SiloedTag` type for the param in `getLogsByTags`. This allowed me to drop some ugly type conversions. This was just a tech debt.

The diff turned out to be quite large because I needed to move the type from `pxe` to `stdlib`.

As mentioned in a PR down the stack this is second of 3 PRs in which I clean up the `getLogsByTags` endpoint:
1. First PR - including block timestamp in return value,
2. This is the second PR - I type the tag arg of the function to be `SiloedTag`,
3. in the last PR I will drop pagination from this endpoint.

benesjan commented:
* **#19290** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19290?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19233** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19125** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19161** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19158** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#19157** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19030** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17611** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17568** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17534** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17445** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17451** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -399,10 +399,7 @@ export class PXEOracleInterface implements ExecutionDataProvider {
         const newLargestIndexMapForIteration: { [k: string]: number } = {};
```
The death of this whole file is imminent so would not pay much attention to this.

spalladino commented on yarn-project/stdlib/src/logs/siloed_tag.ts:
```diff
@@ -1,15 +1,17 @@
  * address that emitted the log.
  */
 export class SiloedTag {
-  private constructor(public readonly value: Fr) {}
+  constructor(public readonly value: Fr) {}
```
I think we have several classes that conform to this interface, maybe you want to add branding to avoid mixing types up? You could even kill this class in favor of re-branding `Fr`s.

benesjan replied:
Introduced branding in https://github.com/AztecProtocol/aztec-packages/pull/19158/commits/7e177d606a0ecb8d977bb977c1c9ba3141477fcc

Thanks!

benesjan commented on yarn-project/archiver/src/archiver/kv_archiver_store/log_store.ts:
```diff
@@ -22,8 +25,12 @@ import type { BlockStore } from './block_store.js';
+  #privateLogsByTag: AztecAsyncMap<string, Buffer[]>;
+  // `{contractAddress}_${tag}` --> public logs
+  #publicLogsByContractAndTag: AztecAsyncMap<string, Buffer[]>;
+  #privateLogKeysByBlock: AztecAsyncMap<number, string[]>;
+  #publicLogKeysByBlock: AztecAsyncMap<number, string[]>;
```
Since in the keys of public logs the key is `{contractAddress}_${tag}` I changed the naming for the private ones here as well and having here `#privateLogKeysByBlock` instead of `#privateLogTagsByBlock`

benesjan commented on yarn-project/archiver/src/archiver/archiver_store_test_suite.ts:
```diff
@@ -2198,46 +2198,35 @@ export function describeArchiverDataStore(
     });
 
-    describe('getLogsByTags', () => {
+    describe('getPrivateLogsByTags', () => {
```
Now the tests were separated to private and public versions.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/noir-structs/log_retrieval_request.ts:
```diff
@@ -9,18 +10,18 @@ import { AztecAddress } from '@aztec/stdlib/aztec-address';
   constructor(
     public contractAddress: AztecAddress,
-    public unsiloedTag: Fr,
+    public tag: Tag,
```
The type contains the info regarding what type of tag it is so made sense to rename this to just "tag".

benesjan commented on yarn-project/pxe/src/tagging/recipient_sync/utils/load_logs_for_range.test.ts:
```diff
@@ -48,48 +46,18 @@ describe('loadLogsForRange', () => {
     expect(await loadLogsForRange(secret, app, aztecNode, 0, 10, NON_INTERFERING_ANCHOR_BLOCK_NUMBER)).toHaveLength(0);
   });
 
-  it('only returns private logs', async () => {
```
The endpoints got split so this test case is no longer necessary 🎉

benesjan commented on yarn-project/pxe/src/logs/log_service.ts:
```diff
@@ -37,16 +35,16 @@ export class LogService {
     return await Promise.all(
       logRetrievalRequests.map(async request => {
-        // TODO(#14555): remove these internal functions and have node endpoints that do this instead
+        // TODO(F-231): remove these internal functions and have node endpoints that do this instead
```
Will tackle https://linear.app/aztec-labs/issue/F-231/update-get-logs-by-tag-node-function-return-value-to-avoid-round-trip in a followup PR to get rid of all the tech debt on the endpoint in one release.

mverzilli replied:
that makes you a really nice person

mverzilli commented on yarn-project/archiver/src/archiver/kv_archiver_store/log_store.ts:
```diff
@@ -100,36 +116,70 @@ export class LogStore {
+      return acc;
+    }, new Map<string, Buffer[]>());
+    const keysOfPrivateLogsToUpdate = Array.from(privateTaggedLogsToAdd.keys());
+    const keysOfPublicLogsToUpdate = Array.from(publicTaggedLogsToAdd.keys());
```
suggestion: split these into auxiliaries to denoise a bit, like (pseudo):

```
const privateTaggedLogsToAdd = this.#extractPrivateTaggedLogs(taggedLogsInBlocks);
const keysOfPrivateLogsToUpdate = Array.from(privateTaggedLogsToAdd.keys());

const publicTaggedLogsToAdd = this.#extractPublicTaggedLogs(taggedLogsInBlocks);
const keysOfPublicLogsToUpdate = Array.from(publicTaggedLogsToAdd.keys());
```

mverzilli replied:
suggestion applies to whole function

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/19158/commits/a63bd730ab8cc1b6d911297e2561e7734d836c6e and in https://github.com/AztecProtocol/aztec-packages/pull/19158/commits/0d693489d65061809fd54e57417033b9b5a0841a

mverzilli commented on yarn-project/archiver/src/archiver/kv_archiver_store/log_store.ts:
```diff
@@ -178,41 +228,82 @@ export class LogStore {
+            return keys ?? [];
+          }),
+        )
+      ).flat();
```
here again I would use some auxiliary functions. when I see it like this I think "there must be a difference" whereas if it was delegating to a `extractTagsToDelete(blocks, this.#privateLogKeysByBlock)` I would know it's the same

benesjan replied:
I was being too lazy here. Cleaned that up in https://github.com/AztecProtocol/aztec-packages/pull/19158/commits/eb61f1ae71e726ff0e2218544e5e468f08d8cc25

Thanks

mverzilli commented on yarn-project/archiver/src/archiver/kv_archiver_store/log_store.ts:
```diff
@@ -178,41 +228,82 @@ export class LogStore {
+   * @param tags - The Tags to filter the logs by.
+   * @param limitPerTag - The maximum number of logs to return per tag.
+   * @returns For each received tag, an array of matching public logs is returned. An empty array implies no logs match
+   * that tag.
```
```suggestion
   * @returns For each received tag, an array of matching public logs is returned. An empty array implies no logs match that tag.
```

mverzilli commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -1408,13 +1416,29 @@ export class Archiver
+   * @param tags - The SiloedTags to filter the logs by.
+   * @param logsPerTag - The number of logs to return per tag. Defaults to everything
+   * @returns For each received tag, an array of matching private logs is returned. An empty array implies no logs match
    * that tag.
```
```suggestion
   * @returns For each received tag, an array of matching private logs is returned. An empty array implies no logs match that tag.
```

mverzilli commented on yarn-project/archiver/src/archiver/archiver_store_test_suite.ts:
```diff
@@ -2253,7 +2242,7 @@ export function describeArchiverDataStore(
           const txEffect = await TxEffect.random();
           txEffect.privateLogs = mockPrivateLogs(blockNumber, txIndex);
-          txEffect.publicLogs = mockPublicLogs(blockNumber, txIndex);
+          txEffect.publicLogs = []; // No public logs needed for private log tests
```
Shouldn't we test some cases with public logs? Even when we know the implementation handles it well now, we probably want to spec that, especially consider the LogStore processes and reduces logs of both types at the same time

benesjan replied:
There is a new block of tests for `getPublicLogsByTagsFromContract`. Or do you think they are insufficient.

mverzilli commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -696,14 +696,29 @@ export class AztecNodeService implements AztecNode, AztecNodeAdmin, Traceable {
+  }
+
+  /**
+   * Gets all public logs that match any of the received tags (i.e. logs with their first field equal to a Tag).
```
In this case probably it's a good idea to add the `contractAddress` param to the docs, to help remark the difference with `getPrivateLogsByTags`

benesjan replied:
Dropped the docs here and left them just on interfaces.

We had too many duplicates and then I forgot to update this one.

https://github.com/AztecProtocol/aztec-packages/pull/19158/commits/e911f98c9c9e0bac95098eaa03f5e2e444eacb56

mverzilli reviewed (approved):
Approving because it's overall good and I won't be around to approve later, but we probably should also add migration notes, shouldn't we?

spalladino reviewed (approved):
Archiver changes look good to me, and agree with @mverzilli on extracting some helpers to make it easier to follow. I also agree with him in that you are a really nice person.

benesjan commented:
> we probably should also add migration notes

Indeed, I have some more changes to this endpoint in the pipeline and I plan on writing them once I am done with those. But thanks for reminding me

benesjan committed:
refactor!: splitting `getLogsByTags` to private and public versions

Fixes #14555

## New PR Description

As mentioned in the old PR description I originally planned to just type the arg of `getLogsByTags` but while addressing this PR's feedback I realized that I made a mistake: the `getLogsByTags` endpoint was used for both private and public logs and the issue is that for public logs we use "unsiloed" tag and for private logs we use siloed tag.

Having these 2 endpoints merged was just a tech debt: there is not a place where we would call this endpoint expecting only 1 type of logs. This led to us having ugly utilities that filtered out one kind of logs after retrieval.

We didn't split this before because we didn't want to introduce a breaking change to the API.

Given that I am already breaking backwards compatibility and given the error I made it made sense to fix this by splitting the two endpoints.

Now we have `getPrivateLogsByTag` and `getPublicLogsByTagsFromContract` endpoints.

## Old PR Description
In this PR I used the `SiloedTag` type for the param in `getLogsByTags`. This allowed me to drop some ugly type conversions. This was just a tech debt.

The diff turned out to be quite large because I needed to move the type from `pxe` to `stdlib`.

As mentioned in a PR down the stack this is second of 3 PRs in which I clean up the `getLogsByTags` endpoint:
1. First PR - including block timestamp in return value,
2. This is the second PR - I type the tag arg of the function to be `SiloedTag`,
3. in the last PR I will drop pagination from this endpoint.

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
